packages/projection/src/operators/withRolledBackEvents.ts

```diff
@@ -0,0 +1,81 @@
+import { ChainSyncEventType, ProjectorOperator, RollBackwardEvent, RollForwardEvent } from '../types';
+import { EMPTY, Observable, map, scan, toArray } from 'rxjs';
+import { WithStabilityWindow } from './withStabilityWindow';
+import { blockingWithLatestFrom } from '@cardano-sdk/util-rxjs';
+
+export type WithRolledBackEvents<ExtraRollForwardProps = {}> = {
+  /**
+   * In reverse order of that they were applied
+   */
+  rolledBackEvents: RollForwardEvent<ExtraRollForwardProps>[];
+};
+
+type WithRolledBackEventsScan<TRollForwardEvent> = {
+  eventCache?: TRollForwardEvent[];
+  evt: TRollForwardEvent | (RollBackwardEvent & WithRolledBackEvents<TRollForwardEvent>);
+};
+
+/**
+ * Adds `rolledBackEvents` to RollBackward events.
+ * `rolledBackEvents` are in descending order (starting from tip going down to origin).
+ *
+ * @param evtCache$ observable that emits events up to first event emitted by source evt$ observable.
+ * It is used to build cache of events to be used in case a rollback happens.
+ * If syncing from origin, there's no need to pass it.
+ * Otherwise, it should emit all events up to source start within stability window.
+ */
+export const withRolledBackEvents =
+  <ExtraRollForwardPropsIn extends WithStabilityWindow, ExtraRollBackwardPropsIn>(
+    evtCache$: Observable<RollForwardEvent<ExtraRollForwardPropsIn>> = EMPTY
+  ): ProjectorOperator<
+    ExtraRollForwardPropsIn,
+    ExtraRollBackwardPropsIn,
+    {},
+    WithRolledBackEvents<RollForwardEvent<ExtraRollForwardPropsIn>>
+  > =>
+  (evt$) =>
+    evt$.pipe(
+      blockingWithLatestFrom(evtCache$.pipe(toArray())),
+      scan(
+        (
+          { eventCache },
+          [evt, initialEvtCache]
+        ): WithRolledBackEventsScan<RollForwardEvent<ExtraRollForwardPropsIn>> => {
+          eventCache ||= initialEvtCache;
+          switch (evt.eventType) {
+            case ChainSyncEventType.RollForward: {
+              // clear blocks that are past stability window
+              const slotThreshold = evt.tip.slot - evt.stabilityWindowSlotsCount;
+              while (eventCache.length > 0 && eventCache[0].block.header.slot < slotThreshold) eventCache.shift();
+              // add current block to cache and return the event unchanged
+              eventCache.push(evt);
+              return { eventCache, evt };
+            }
+            case ChainSyncEventType.RollBackward: {
+              const rollbackTo = evt.tip;
+              if (rollbackTo === 'origin') {
+                // Review: consider checking if `eventCache` isn't missing some blocks due to
+                // not syncing from genesis AND not providing sufficient evtCache$.
+                // Could be nice to have, but also could be redundant
+                // if we test our implementation of creating evtCache$
```

Comment:
Maybe just add this then?

Reply:
Done in 03cfe9408d1f2e37273a33c921e91e2c3e6b9455